### PR TITLE
Test add capacity with restart node, when capacity near full ratio

### PR DESCRIPTION
**Pull request by yitzhak12**
- Use monitoring to fill up the cluster, and create a new fixture in 'monitoring/conftest'
that will help with storageutilization of rbd, and cephfs.
- Perform IO using the method 'run_io' of the pod object.
- Make use in the functions that written in the branch 'WIP-clexp-entry'.
After this branch will be merged into master I will use them directly.
- Start one or more of the worker nodes when add_capacity is executed.

### CODE REVIEW

File: tests/manage/monitoring/conftest.py [hunk 1]
```diff
@@ -566,6 +566,92 @@ def workload_storageutilization_10g_cephfs(
     return measured_op
 
 
+@pytest.fixture
```
Comment (ebenahar) [inline tests/manage/monitoring/conftest.py:569]: are the changes in this file intentional?

File: tests/manage/z_cluster/cluster_expansion/temp_helper_file.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
+import logging
```
Comment (ebenahar) [inline tests/manage/z_cluster/cluster_expansion/temp_helper_file.py:1]: The functions in this file should be moved to the corresponding files under ocs_ci/

File: tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py [hunk 3]
```diff
@@ -0,0 +1,202 @@
+import pytest
+import logging
+import time
+from tests.manage.monitoring.conftest import *
```
Comment (ebenahar) [inline tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py:4]: no need to import from conftest. It's being done automatically by pytest.
Also, in general, we should not import from one test module to another.

File: tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+
+def check_osd_pods_after_expansion(osd_pods_before):
+    osd_pods_after = pod_helpers.get_osd_pods()
+    number_of_osds_added = len(osd_pods_after) - len(osd_pods_before)
+    logging.info(f"number of osd's added = {number_of_osds_added}, "
+                 f"before = {len(osd_pods_before)}, after = {len(osd_pods_after)}")
+    if number_of_osds_added != 3:
+        return False
+
+    return True
```
Comment (ebenahar) [inline tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py:36]: please move these functions to the relevant files under ocs_ci/

File: tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+
+
+def check_total_space_after_expansion(total_space_b4_expansion):
```
Comment (ebenahar) [inline tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py:39]: change `b4` to `before`

File: tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+from ocs_ci.ocs import constants
+from ocs_ci.ocs.ocp import OCP
+from ocs_ci.ocs.resources import pod as pod_helpers
+from ocs_ci.ocs.resources import storage_cluster
+from ocs_ci.utility.utils import ceph_health_check
+from ocs_ci.ocs.node import get_typed_nodes, wait_for_nodes_status
+from tests.manage.z_cluster.cluster_expansion import temp_helper_file
+
+
+def get_total_space():
```
Comment (ebenahar) [inline tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py:19]: please add docstring to the functions in this file

File: tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+        """
+        logging.info("Condition 1 to start the test is met: storageutilization is completed")
+        used_capacity_after_utilizing = temp_helper_file.get_percent_used_capacity()
+        logging.info(f"### used capacity after storageutilization = {used_capacity_after_utilizing}")
+
+        seconds_to_wait_after_storageutilization = 60
+        logging.info(f"wait {seconds_to_wait_after_storageutilization} seconds before starting the test")
+        time.sleep(seconds_to_wait_after_storageutilization)
+
+        assert temp_helper_file.check_pods_in_running_state(), \
```
Comment (ebenahar) [inline tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py:125]: please break lines with parentheses, as follows
```python
assert temp_helper_file.check_pods_in_running_state(), (
    "Condition 2 to start test failed: "
    "one or more OCS pods are not in running state"
)
```

Comment (ebenahar) [reply tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py:125]: Please change through

File: tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+        else:
+            return temp_helper_file.check_osd_tree_3az_aws(tree_output, len(osd_pods))
+
+
+@pytest.mark.parametrize(
+    argnames=["node_type", "num_of_nodes"],
+    argvalues=[
+        pytest.param(
+            *['worker', 1],
+            marks=pytest.mark.polarion_id("OCS-1311")
```
Comment (hnallurv) [inline tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py:85]: OCS-1311: OCS monitoring includes volume usage in bytes from CSI plugin for RBD/XFS backed volume. This test is not related to this PR
It should be OCS-1313. Please change.

File: tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+@tier4
+@polarion_id('OCS-1313')
+class TestAddCapacityNodeRestart(ManageTest):
+    """
+    Test add capacity when one of the nodes got restart
+    in the middle of the process. Don't forget to include the flag '--cluster-name' when running the test
+    """
+    num_of_pvcs = 4
+    pvc_size = 100
+
```
Comment (hnallurv) [inline tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py:109]: Looks like this test is not running any background IOs. As per test case OCS-1313, "Make sure this test has the configuratio CFG#1 in https://docs.google.com/spreadsheets/d/1yMzw6SzWwcrB1nM5atEVVNc9FbpRV9cOEUDPlteMQoo/edit#gid=978210857&range=C3 and Backgroud operatios as mentioned in https://docs.google.com/spreadsheets/d/1yMzw6SzWwcrB1nM5atEVVNc9FbpRV9cOEUDPlteMQoo/edit#gid=978210857&range=B17". Please try to follow the entry criteria mentioned there.

File: tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+        pod_objs = []
+
+        with ThreadPoolExecutor(max_workers=self.num_of_pvcs) as executor:
+            for pvc_obj in pvc_objs:
+                executor.submit(pod_objs.append, (pod_factory(pvc=pvc_obj)))
+
+        with ThreadPoolExecutor(max_workers=self.num_of_pvcs-1) as executor:
+            for pod_obj in pod_objs:
+                pod_io_task = executor.submit(
+                    pod_obj.run_io, storage_type='fs', size=4, jobs=10, io_direction='wo', rate='250m')
```
Comment (hnallurv) [inline tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py:154]: Will these IOs keep running throughout the test?

File: tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+
```
Comment (hnallurv) [inline tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py:155]: Need to cover NooBaa IOs also....

File: tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+        if not check_total_space_after_expansion(total_space_b4_expansion):
+            logging.warning("Expected capacity mismatch")
+        if not check_osd_tree():
+            logging.warning("Incorrect ceph osd tree format")
+
+        logging.info("Finished verifying add capacity osd storage with node restart")
+        logging.info("Waiting for ceph health check to finished...")
+        ceph_health_check(
+            namespace=config.ENV_DATA['cluster_namespace'], tries=60
+        )
```
Comment (hnallurv) [inline tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py:202]: I suggest referring to PR https://github.com/red-hat-storage/ocs-ci/pull/1751. It's almost covering all the entry and exit criteria. All you need is to add failure condition and verification items. It's better to re-use the code from there for this test and also other Tier4 tests.

File: tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+from ocs_ci.ocs.resources import pod as pod_helpers
+from ocs_ci.ocs.resources import storage_cluster
+from ocs_ci.utility.utils import ceph_health_check
+from ocs_ci.ocs.node import get_typed_nodes, wait_for_nodes_status
+
+
+# The functions below I took from the branch 'wip-add-capacity-e_e'.
+# When this branch will be merged into master I will use the functions from this branch.
+
+def get_percent_used_capacity():
```
Comment (ebenahar) [inline tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py:20]: please move this function to `ocs_ci/ocs/cluster.py`

File: tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+        float: The percentage of the used capacity in the cluster
+    """
+    ct_pod = pod_helpers.get_ceph_tools_pod()
+    output = ct_pod.exec_ceph_cmd(ceph_cmd='ceph df')
+    total_used = (output.get('stats').get('total_used_raw_bytes'))
+    total_avail = (output.get('stats').get('total_bytes'))
+    return 100.0 * total_used / total_avail
+
+
+def check_pods_in_running_state(namespace=defaults.ROOK_CLUSTER_NAMESPACE):
```
Comment (ebenahar) [inline tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py:34]: please move this function to `ocs_ci/ocs/resources/pod.py`

File: ocs_ci/ocs/fiojob.py [hunk 15]
```diff
@@ -548,6 +551,10 @@ def is_storage_reclaimed():
         # Without checksum, we just need to make sure that data were deleted
         # and wait for this to happen to avoid conflicts with tests executed
         # right after this one.
-        delete_fio_data(fio_job_file, is_storage_reclaimed)
+        if delete_fio_data_after_test:
+            delete_fio_data(fio_job_file, is_storage_reclaimed)
+        # Wait until the end of the tests, and then delete the data.
+        else:
+            request.addfinalizer(lambda: delete_fio_data(fio_job_file, is_storage_reclaimed))
```
Comment (ebenahar) [inline ocs_ci/ocs/fiojob.py:558]: I haven't seen `addfinalizer` being used not in fixtures before. Is this a correct usage?

File: tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+
+# The functions below I took from the branch 'wip-add-capacity-e_e'.
+# When this branch will be merged into master I will use the functions from this branch.
+
+def get_percent_used_capacity():
+    """
+    Function to calculate the percentage of used capacity in a cluster
+
+    Returns:
+        float: The percentage of the used capacity in the cluster
```
Comment (ebenahar) [inline tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py:25]: add an empty line below

File: tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py [hunk 17]
```diff
@@ -0,0 +1,169 @@
...
+    ct_pod = pod_helpers.get_ceph_tools_pod()
+    output = ct_pod.exec_ceph_cmd(ceph_cmd='ceph df')
+    total_used = (output.get('stats').get('total_used_raw_bytes'))
+    total_avail = (output.get('stats').get('total_bytes'))
+    return 100.0 * total_used / total_avail
+
+
+def check_pods_in_running_state(namespace=defaults.ROOK_CLUSTER_NAMESPACE):
+    """
+    checks whether all the pods in a given namespace are in Running state or not
```
Comment (ebenahar) [inline tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py:36]: Change to: `Checks`

Comment (ebenahar) [inline tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py:36]: add an empty line below

File: tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py [hunk 18]
```diff
@@ -0,0 +1,169 @@
...
+    Returns:
+        Boolean: True, if all pods in Running state. False, otherwise
```
Comment (ebenahar) [inline tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py:38]: add an empty line below

File: tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py [hunk 19]
```diff
@@ -0,0 +1,169 @@
...
+            marks=pytest.mark.polarion_id("OCS-1313")
+        ),
+     ]
+)
+@pytest.mark.parametrize(
+    "workload_storageutilization_rbd",
+    [(0.25, False, 120)],
+    indirect=["workload_storageutilization_rbd"]
+)
+# @pytest.mark.parametrize(
```
Comment (ebenahar) [inline tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py:67]: should this be commented out?

File: tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py [hunk 20]
```diff
@@ -0,0 +1,169 @@
...
+#     "workload_storageutilization_cephfs",
+#     [(0.05, False, 120)],
+#     indirect=["workload_storageutilization_cephfs"]
+# )
+@ignore_leftovers
+@tier4
+@polarion_id('OCS-1313')
```
Comment (ebenahar) [inline tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py:74]: not needed as you have this in line #58

File: tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py [hunk 21]
```diff
@@ -0,0 +1,169 @@
...
+    """
+    num_of_pvcs = 3
+    pvc_size = 100
+
+    def test_add_capacity_node_restart(
+        self, nodes, multi_pvc_factory, pvc_factory, pod_factory, workload_storageutilization_rbd,
+        node_type, num_of_nodes,
+    ):
+        """ test add capacity when one of the nodes got restart
+        in the middle of the process
```
Comment (ebenahar) [inline tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py:88]: change to:
```
""" 
test add capacity when one of the nodes got restart ￼in the middle of the process
"""

File: tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py [hunk 22]
```diff
@@ -0,0 +1,169 @@
...
+        """
+        logging.info("Condition 1 to start the test is met: storageutilization is completed")
+        used_capacity_after_utilizing = get_percent_used_capacity()
+        logging.info(f"### used capacity after storageutilization = {used_capacity_after_utilizing}")
```
Comment (ebenahar) [inline tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py:92]: delete the `###`

File: tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py [hunk 23]
```diff
@@ -0,0 +1,169 @@
...
+
+        seconds_to_wait_after_storageutilization = 60
```
Comment (ebenahar) [inline tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py:94]: please don't wait for an absolute number of seconds. Instead create a functionality to poll what you need.
You can use `ocs_ci.utility.utils.TimeoutSampler` or ocs_ci.utility.retry.retry

File: ocs_ci/ocs/fiojob.py (hunk 15 above)
Comment (yitzhak12) [reply ocs_ci/ocs/fiojob.py:558]: Here are 2 examples that I can see in 'tests/conftest' file: 
https://github.com/red-hat-storage/ocs-ci/blob/master/tests/conftest.py#L833
https://github.com/red-hat-storage/ocs-ci/blob/master/tests/conftest.py#L901

Comment (ebenahar) [reply ocs_ci/ocs/fiojob.py:558]: ack, thanks for checking

File: tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py (hunk 11 above)
Comment (ebenahar) [reply tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py:77]: @hnallurv maybe in a follow up PR? this one is already quite large

File: tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py [hunk 24]
```diff
@@ -0,0 +1,114 @@
...
+        pytest.param(
+            *['worker', 1],
+            marks=pytest.mark.polarion_id("OCS-1313")
+        ),
+     ]
+)
+@pytest.mark.parametrize(
+    "workload_storageutilization_rbd",
+    [(0.25, False, 120)],
+    indirect=["workload_storageutilization_rbd"]
```
Comment (ebenahar) [inline tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py:28]: need to add a polarion_id here as well

File: tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py [hunk 25]
```diff
@@ -0,0 +1,117 @@
...
+            *['worker', 1],
+            marks=pytest.mark.polarion_id("OCS-1313")
+        ),
+     ]
+)
+@pytest.mark.parametrize(
+    argnames=["workload_storageutilization_rbd"],
+    argvalues=[pytest.param(
+        (0.25, False, 120),
+        marks=pytest.mark.polarion_id("OCS-1313"),
```
Comment (ebenahar) [inline tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py:29]: it's the same polarion ID as in line21. Can you please create a separate Polarion test case?

File: ocs_ci/ocs/fiojob.py (hunk 15 above)
Comment (mbukatov) [inline ocs_ci/ocs/fiojob.py:558]: Let me think whether this is a good idea or not.

File: tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py (hunk 9 above)
Comment (yitzhak12) [reply tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py:35]: As I wrote in my comment here: https://github.com/red-hat-storage/ocs-ci/pull/1996/files#diff-b451e3caa51bd819db2305fc5466e846R44 these changes will be replaced when the branch 'wip-add-capacity-e_e' will be merged into master.

File: tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py (hunk 10 above)
Comment (yitzhak12) [reply tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py]: The IOs here are not complete. When the branch 'wip-add-capacity-e_e' will be merged into master I will use the functions from this branch. Currently, I think there is no need to reinvent the wheel and we don't want to wait for this PR until branch 'wip-add-capacity-e_e' will be merged into master, cause it may take some time.

File: tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py (hunk 11 above)
Comment (yitzhak12) [reply tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py:73]: The IOs here are not complete. When the branch 'wip-add-capacity-e_e' will be merged into master I will use the functions from this branch. Currently, I think there is no need to reinvent the wheel and we don't want to wait for this PR until branch 'wip-add-capacity-e_e' will be merged into master, cause it may take some time.

File: tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py (hunk 12 above)
Comment (yitzhak12) [reply tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py:110]: When the branch 'wip-add-capacity-e_e' will be merged into master I will use the exit criteria from this branch. Currently, we don't want to wait for this PR will be merged into master, cause it may take some time.

File: tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py (hunk 8 above)
Comment (yitzhak12) [reply tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py]: I changed it to OCS-1313

File: tests/manage/z_cluster/conftest.py [hunk 26]
```diff
@@ -0,0 +1,97 @@
...
+            For example: Let's say I want to use workload_storageutilization_cephfs fixture with
+            'target_percentage'=0.25, 'delete_fio_job_file'=True, 'minimal_time'=120
+            then In my test I will specify these parameters:
+            @pytest.mark.parametrize("workload_storageutilization_cephfs",
+            [(0.25, False, 120)], indirect=["workload_storageutilization_cephfs"])
+            """
+
+    target_percentage, delete_fio_data_after_test, minimal_time = request.param
+    percent_to_fill = int(target_percentage*100)
+    fixture_name = "workload_storageutilization_{}p_cephfs".format(str(percent_to_fill))
```
Comment (fbalak) [inline tests/manage/z_cluster/conftest.py:83]: There should be used f-string instead of format

File: tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py [hunk 27]
```diff
@@ -0,0 +1,110 @@
...
+class TestAddCapacityNodeRestart(ManageTest):
+    """
+    Test add capacity when one of the nodes got restart
+    in the middle of the process. Don't forget to include the flag '--cluster-name' when running the test
+    """
+    num_of_pvcs = 3
+    pvc_size = 100
+
+    def test_add_capacity_node_restart(
+        self, nodes, multi_pvc_factory, pvc_factory, pod_factory, workload_storageutilization_rbd,
```
Comment (fbalak) [inline tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py:37]: `pvc_factory` fixture is not used in test

File: ocs_ci/ocs/fiojob.py [hunk 28]
```diff
@@ -299,6 +299,9 @@ def workload_fio_storageutilization(
     target_percentage=None,
     target_size=None,
     with_checksum=False,
+    delete_fio_data_after_test=True,
```
Comment (fbalak) [inline ocs_ci/ocs/fiojob.py:302]: Please document new parameters in docstring

**Comment by yitzhak12** [comment]
@fbalak is it ok now, or need to change something?

File: ocs_ci/ocs/fiojob.py [hunk 29]
```diff
@@ -339,6 +342,13 @@ def workload_fio_storageutilization(
             fio is stored on the volume, and reclaim policy of the volume is
             changed to ``Retain`` so that the volume is not removed during test
             teardown for later verification runs
+        delete_fio_data_after_test (bool): If true, delete fio data after the fio
+            storageutilization is completed. Else if False, delete the fio data at
+            the end of all the tests. If if false you also need to provide the
+            request parameter.
+        request: The built-in pytest request object.
```
Comment (fbalak) [inline ocs_ci/ocs/fiojob.py:349]: I would write that this is required if `delete_fio_data_after_test` is set to false.

File: ocs_ci/ocs/fiojob.py [hunk 30]
```diff
@@ -548,6 +558,10 @@ def is_storage_reclaimed():
         # Without checksum, we just need to make sure that data were deleted
         # and wait for this to happen to avoid conflicts with tests executed
         # right after this one.
-        delete_fio_data(fio_job_file, is_storage_reclaimed)
+        if delete_fio_data_after_test:
+            delete_fio_data(fio_job_file, is_storage_reclaimed)
+        # Wait until the end of the tests, and then delete the data.
+        else:
+            request.addfinalizer(lambda: delete_fio_data(fio_job_file, is_storage_reclaimed))
```
Comment (fbalak) [inline ocs_ci/ocs/fiojob.py:565]: It might be good to add log info that `request` object have to be provided as parameter. It might help to debug for people that use this function without providing `request` object which has default value set to `None`.

File: ocs_ci/ocs/fiojob.py [hunk 31]
```diff
@@ -545,9 +556,14 @@ def is_storage_reclaimed():
...
         ocp_pv.add_label(pv_name, label)
     else:
-        # Without checksum, we just need to make sure that data were deleted
-        # and wait for this to happen to avoid conflicts with tests executed
+        # Without checksum, we just need to make sure that the data will be deleted.
+        # If 'delete_fio_data_after_test' is False and 'request' object is provided
+        # Wait until the end of the tests, and then delete the data.
+        if not delete_fio_data_after_test and request:
+            request.addfinalizer(lambda: delete_fio_data(fio_job_file, is_storage_reclaimed))
+        # Wait for the data to be deleted, to avoid conflicts with tests executed
```
Comment (mbukatov) [inline ocs_ci/ocs/fiojob.py:564]: This change obstructs the original meaning a bit.

Comment (yitzhak12) [reply ocs_ci/ocs/fiojob.py:564]: ok, I will change it to what it was before. I really wasn't sure about that myself. I am just thinking about what we should do if the user didn't provide the 'request' param.

File: ocs_ci/ocs/fiojob.py (hunk 15 above)
Comment (mbukatov) [reply ocs_ci/ocs/fiojob.py:567]: Ok, so we have 2 options here:

- drop both `delete_fio_data_after_test` and `request` options, and use something like `keep_fio_data` option instead, so that when this is True, `delete_fio_data()` is not executed
- move execution of  `delete_fio_data()` to teardown phase as done now, but simplify the interface (instead of  `delete_fio_data_after_test` and `request`, use single option `teardown_request` with explanation that when you pass a pytest request object there, the data will be explicitly deleted during test case teardown. But for this to work in all cases, we need to make sure that `max_avail_before_delete` is also populated during teardown.

Otherwise I like this.

File: ocs_ci/ocs/fiojob.py [hunk 32]
```diff
@@ -548,6 +554,9 @@ def is_storage_reclaimed():
         # Without checksum, we just need to make sure that data were deleted
         # and wait for this to happen to avoid conflicts with tests executed
         # right after this one.
-        delete_fio_data(fio_job_file, is_storage_reclaimed)
+        if not keep_fio_data:
+            delete_fio_data(fio_job_file, is_storage_reclaimed)
+        else:
+            logging.info("The fio data will be deleted at the teardown part")
```
Comment (mbukatov) [inline ocs_ci/ocs/fiojob.py:560]: Minor suggestion to make it more clear why and when the data are actually deleted in this case:
```suggestion
            logging.info("The fio data will be deleted during project teardown")
```

File: ocs_ci/ocs/fiojob.py (hunk 15 above)
Comment (mbukatov) [reply ocs_ci/ocs/fiojob.py]: I discussed this with Itzhak and we agreed to with 1st option, because it seems to work good enough for this new usecase and requires minimal change for this PR to work.

File: tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py [hunk 33]
```diff
@@ -0,0 +1,111 @@
...
+@pytest.mark.parametrize(
+    argnames=["node_type", "num_of_nodes", "workload_storageutilization_rbd"],
+    argvalues=[
+        pytest.param(
+            *['worker', 1, (0.11, True, 120)],
+            marks=pytest.mark.polarion_id("OCS-1313")
+        ),
+    ],
+    indirect=["workload_storageutilization_rbd"],
+)
```
Comment (ebenahar) [inline tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py:25]: there is only a single set of params here. Is pytest.mark.parametrize even needed?

File: tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py [hunk 34]
```diff
@@ -0,0 +1,111 @@
...
+@ignore_leftovers
+@tier4
```
Comment (ebenahar) [inline tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py:27]: need to mark this is tier4a / tier4a / tier4b. Please pick tier4a

File: tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py (hunk 33 above)
Comment (yitzhak12) [reply tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py:25]: I think it will be easier to write it that way. And also we need it for the pytest indirect parameters, as I can see in the docs: https://docs.pytest.org/en/latest/example/parametrize.html#apply-indirect-on-particular-arguments

Comment (ebenahar) [reply tests/manage/z_cluster/cluster_expansion/test_add_capacity_with_node_restart.py:25]: ok

**Comment by yitzhak12** [comment]
@ebenahar please marked as verified
